File: src/repositories/vendor.repository.ts
import { Collection, ObjectId } from "mongodb";
import type { Vendor } from "@/types/vendor/vendor";
import { logger } from "@/utils/logger";
import { getCompanyVendorsCollection, getVendorsCollection } from "@/repositories/collections/core.collections";

export class VendorRepository {
  private getCollection(companyId?: string): Collection<Vendor> {
    if (companyId) {
      return getCompanyVendorsCollection(companyId);
    }
    return getVendorsCollection();
  }

  async create(vendor: Omit<Vendor, "_id">, companyId?: string): Promise<Vendor> {
    try {
      const collection = this.getCollection(companyId);
      const result = await collection.insertOne(vendor as any);
      const created = { ...vendor, _id: result.insertedId } as Vendor;
      logger.info("Vendor created", { vendorId: result.insertedId, companyId });
      return created;
    } catch (error) {
      logger.error("Failed to create vendor", error);
      throw error;
    }
  }

  async findById(id: string, companyId?: string): Promise<Vendor | null> {
    try {
      const collection = this.getCollection(companyId);
      return await collection.findOne({ _id: new ObjectId(id) });
    } catch (error) {
      logger.error("Failed to find vendor by ID", error);
      throw error;
    }
  }

  async findAll(companyId?: string): Promise<Vendor[]> {
    try {
      return await this.getCollection(companyId)
        .find({})
        .sort({ name: 1 })
        .toArray();
    } catch (error) {
      logger.error("Failed to fetch all vendors", error);
      throw error;
    }
  }

  async findByIds(ids: string[], companyId?: string): Promise<Vendor[]> {
    try {
      const collection = this.getCollection(companyId);
      const objectIds = ids.map(id => new ObjectId(id));
      return await collection
        .find({ _id: { $in: objectIds } })
        .sort({ name: 1 })
        .toArray();
    } catch (error) {
      logger.error("Failed to fetch vendors by IDs", error);
      throw error;
    }
  }

  async update(id: string, updates: Partial<Vendor>, companyId?: string): Promise<Vendor | null> {
    try {
      const collection = this.getCollection(companyId);
      return await collection.findOneAndUpdate(
        { _id: new ObjectId(id) },
        { $set: updates },
        { returnDocument: "after" }
      );
    } catch (error) {
      logger.error("Failed to update vendor", error);
      throw error;
    }
  }

  async delete(id: string, companyId?: string): Promise<boolean> {
    try {
      const collection = this.getCollection(companyId);
      const result = await collection.deleteOne({ _id: new ObjectId(id) });
      const deleted = result.deletedCount > 0;

      if (deleted) {
        logger.info("Vendor deleted", { vendorId: id, companyId });
      }

      return deleted;
    } catch (error) {
      logger.error("Failed to delete vendor", error);
      throw error;
    }
  }

  async exists(id: string, companyId?: string): Promise<boolean> {
    try {
      const collection = this.getCollection(companyId);
      const count = await collection.countDocuments({ _id: new ObjectId(id) });
      return count > 0;
    } catch (error) {
      logger.error("Failed to check vendor existence", error);
      throw error;
    }
  }
}

File: fisoft/src/main/kotlin/repository/VendorRepository.kt
package com.ilyasdemirkiran.repository

import com.ilyasdemirkiran.config.Database
import com.ilyasdemirkiran.types.vendors.Vendor
import com.mongodb.client.model.Filters
import com.mongodb.client.model.FindOneAndUpdateOptions
import com.mongodb.client.model.ReturnDocument
import com.mongodb.client.model.Updates
import com.mongodb.kotlin.client.coroutine.MongoCollection
import kotlinx.coroutines.flow.firstOrNull
import kotlinx.coroutines.flow.toList
import org.bson.Document
import org.bson.types.ObjectId

class VendorRepository {
    private fun getCollection(): MongoCollection<Vendor> {
        return Database.getGlobalVendorDatabase().getCollection<Vendor>("vendors")
    }

    suspend fun create(vendor: Vendor): Vendor {
        val id = vendor._id ?: ObjectId().toHexString()
        val created = vendor.copy(_id = id)
        getCollection().insertOne(created)
        return created
    }

    suspend fun findById(id: String): Vendor? {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        return getCollection().find(filter).firstOrNull()
    }

    suspend fun findAll(): List<Vendor> {
        return getCollection().find()
            .sort(Document("name", 1))
            .toList()
    }

    suspend fun findByIds(ids: List<String>): List<Vendor> {
        val objIds = ids.map { try { ObjectId(it) } catch (e: Exception) { it } }
        return getCollection().find(Filters.`in`("_id", objIds))
            .sort(Document("name", 1))
            .toList()
    }

    suspend fun update(id: String, name: String? = null, phone: String? = null, city: String? = null, district: String? = null, address: String? = null): Vendor? {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        val updates = mutableListOf<org.bson.conversions.Bson>()
        if (name != null) updates.add(Updates.set("name", name))
        if (phone != null) updates.add(Updates.set("phone", phone))
        if (city != null) updates.add(Updates.set("city", city))
        if (district != null) updates.add(Updates.set("district", district))
        if (address != null) updates.add(Updates.set("address", address))

        if (updates.isEmpty()) return findById(id)

        return getCollection().findOneAndUpdate(
            filter,
            Updates.combine(updates),
            FindOneAndUpdateOptions().returnDocument(ReturnDocument.AFTER)
        )
    }

    suspend fun delete(id: String): Boolean {
        val filter = try { Filters.eq("_id", ObjectId(id)) } catch (e: Exception) { Filters.eq("_id", id) }
        val res = getCollection().deleteOne(filter)
        return res.deletedCount > 0
    }
}
